### PR TITLE
Add Sundials to the JuliaDiffEq common interface

### PR DESCRIPTION
**Pull request by ChrisRackauckas**
This is an extension of the previous PR #87

The common interface is a type-based dispatch interface which allows all of the JuliaDiffEq solvers to be used uniformly. It is being discussed here:

https://github.com/JuliaDiffEq/Roadmap/issues/5

Usage of the interface is summarized as follows. One defines a problem `prob = ODEProblem(f,u0,tspan)` and then the common interface is a set of dispatches on `solve`:

```julia
sol = solve(prob,alg;kwargs...)
```

were there's an agreed upon comprehensive list of keyword arguments which will be documented in DiffEqDocs.jl. The resulting solve object feels like an array:

```julia
sol[i] # the ith solution value
sol.t[i] # the ith timepoint
```

and I will soon add the interpolation part of the interface

```julia
sol(t) # the interpolated solution at time t
```

By joining the common interface, all of the tools in the DiffEq universe can use Sundials as a backend: parameter estimation, benchmarking, etc. 

This will look very familiar because it's based on the interface that used to live in OrdinaryDiffEq.jl (in DifferentialEquations.jl), but is being distributed amongst the separate packages (PR49, ODEInterface, Sundials, etc.) for easier maitenance and to show contributors "how to plug into the ecosystem". 

The result is a unified interface with modular implementations so that way it's easy to maintain and extend.

----------------------

This PR is listed as WIP since the common interface is still not completely solid. However, most of this is already together so please start review. Thanks!

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
This currently requires the `common_interface` branch of DiffEqBase

**Comment by ChrisRackauckas** [comment]
The JuliaDiffEq ecosystem tags are in metadata right now which will be required for tests to pass.

Question: should this be a separate package SundialsDiffEq.jl to separate it from the Sundials wrapper itself?

**Comment by mauro3** [comment]
It should be just one package, I see no use in splitting it up as the amount of code is not big.

**Comment by ChrisRackauckas** [comment]
I agree. I can't see this getting larger than 300 lines of code, even with the full common interface implemented. Note that this is a pretty bare version of the interface: more details can be added in later.

**Comment by mauro3** [comment]
Solver-packages in JuliaDiffEq should contain the warppers.  I don't think it is good for the user to always have to include package+wrapper-package.  For packages outside, we should encourage them to add the wrapper to the package.  If that doesn't happen we can a wrapper to JuliaDiffEq.

**Comment by ChrisRackauckas** [comment]
>I don't think it is good for the user to always have to include package+wrapper-package.

Well they usually wouldn't need to if they use the metapackage which I'd recommend, but yes it should be the easiest way possible. I think that this setup makes more sense for PyDSTool.jl + DiffEqBifurcate.jl  where PyDSTool will likely end up being one backend to DiffEqBifurcate.jl, and thus should be sequestered in some sense.

>For packages outside, we should encourage them to add the wrapper to the package. If that doesn't happen we can a wrapper to JuliaDiffEq.

I am trying to get in contact with the developer for ODEInterface about this... no luck so far. I still have the conditional dependency code in OrdinaryDiffEq and would like to get rid of the conditional deps.

File: src/simple.jl [hunk 1]
```diff
@@ -129,65 +129,6 @@ function cvode(f::Function, y0::Vector{Float64}, t::Vector{Float64}, userdata::A
...
-* `tspan`, a vector where `tspan[1]` is the starting time and the other values are
-  time values which are guaranteed in the output
-* `integrator`, the chosen integration algorithm. Default is `:BDF`
-  , other option is `:Adams`
-* `reltol`, Relative Tolerance to be used (default=1e-3)
-* `abstol`, Absolute Tolerance to be used (default=1e-6)
-
-return: a vector with the timepoints `t` and a vector with the outputs `y0`
-"""
-function cvode_fulloutput(f::Function, y0::Vector{Float64}, tspan::Vector{Float64}, userdata::Any = nothing;
```
Comment (musm) [inline src/simple.jl]: doesn't this need a depwarn?

File: src/common.jl [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+                          u = vec(u); du=vec(du); 0)
+    end
+
+    if alg == CVODE_BDF
+        mem = CVodeCreate(CV_BDF, CV_NEWTON)
+    elseif alg ==  CVODE_Adams
+        mem = CVodeCreate(CV_ADAMS, CV_FUNCTIONAL)
+    end
+
+    if mem == C_NULL
```
Comment (musm) [inline src/common.jl]: `mem` doesn't look initialized so if alg is not CVODE_BDF or CVODE_Adams, this could be problematic?

Comment (ChrisRackauckas) [inline src/common.jl]: it can only dispatch to this method if it's either of those algorithms.

**Comment by ChrisRackauckas** [comment]
Added the depwarn printing and put an automatic conversion

**Comment by ChrisRackauckas** [comment]
Moved as requested

File: src/common.jl [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+                    flag = @checkflag CVodeGetDky(
+                                            mem, save_ts[k], Cint(0), ures[end])
+                    ts[end] = save_ts[k]
+                else # Just push another value
+                    flag = @checkflag CVodeGetDky(
+                                            mem, save_ts[k], Cint(0), ures[end])
+                    push!(ures,copy(u))
+                    push!(ts, save_ts[k]...)
+                end
+        end
```
Comment (musm) [inline src/common.jl]: indent is off

Comment (ChrisRackauckas) [inline src/common.jl]: Fixed

**Comment by musm** [comment]
so does Sundails now need a new tag due to the exported interface change?

**Comment by ChrisRackauckas** [comment]
It just went through earlier today.

**Pull request merged**